spec/OneClickInstall_spec.rb

```diff
@@ -28,36 +28,170 @@
+      repos = subject.GetNonRequiredRepositories
+      expect(repos).to match_array(urls[1])
+    end
+  end
+
+  describe "#GetRequiredSoftware" do
+    it "Returns a list of the names of the software currently selected for installation" do
+      software_to_install = subject.GetRequiredSoftware
+      expect(software_to_install).to match_array(["vim" , "x11"])
+    end
+  end
+
+  describe "#GetRequiredRemoveSoftware" do
+    it "Returns a list of the names of the software currently selected for removal" do
+      software_to_remove = subject.GetRequiredRemoveSoftware
+      expect(software_to_remove).to match_array(["sap_server"])
+    end
+  end
+
+  describe "#GetRequiredPackages" do
+    it "Returns a list of the names of the packages currently selected for installation" do
+      packages_to_install = subject.GetRequiredPackages
+      expect(packages_to_install).to match_array(["vim"])
+    end
+  end
+
+  describe "#GetRequiredPatterns" do
+    it "Returns a list of the names of the patterns currently selected for installation" do
+      patterns_to_install = subject.GetRequiredPatterns
+      expect(patterns_to_install).to match_array(["x11"])
+    end
+  end
+
+  describe "#HaveAnyRecommended" do
+    it "Returns boolean, depending on the existence of any recommended repositories or software" do
+      any_recommended = subject.HaveAnyRecommended
+      expect(any_recommended).to be true
+    end
+  end  
+
+  describe "#makeXMLFriendly" do
+    it "converts map structure (eg. @repositories) to a list of maps with a key element" do
+      flattened = subject.makeXMLFriendly(software)
+      expect(flattened).to match_array([
+        {"recommended"=>"true", "action"=>"remove", "type"=>"pattern", "key"=>"sap_server"},
+        {"recommended"=>"false", "action"=>"install", "type"=>"package", "key"=>"tftp"},
+        {"recommended"=>"true", "action"=>"install", "type"=>"package", "key"=>"vim"},
+        {"recommended"=>"true", "action"=>"install", "type"=>"pattern", "key"=>"x11"}])
+    end
+  end
+
+  describe "#fromXMLFriendly" do
+    it "converts back from makeXMLFriendly to original structure" do
+      flattened = [
+        {"recommended"=>"true", "action"=>"remove", "type"=>"pattern", "key"=>"sap_server"},
+        {"recommended"=>"false", "action"=>"install", "type"=>"package", "key"=>"tftp"},
+        {"recommended"=>"true", "action"=>"install", "type"=>"package", "key"=>"vim"},
+        {"recommended"=>"true", "action"=>"install", "type"=>"pattern", "key"=>"x11"}]
+      unflattened = subject.fromXMLFriendly(flattened)
+      expect(unflattened).to match_array(software)
+    end
+  end
+
+  describe "#ToXML" do
+    it "serializes to XML file, according to OneClickInstall data structure" do
+      filename = File.join(DATA_PATH, "serialized_result.xml")
+      subject.ToXML(filename)
+      expected = File.join(DATA_PATH, "serialized_expected.xml") 
+      expect(IO.read(filename)).to eql(IO.read(expected))
```

Comment:
Just note that can be quite fragile as identical XML is not necessary same as same content of xml due to formatting. Maybe comparing parsed xml trees is better way?

Reply:
I tried to replace it with `expect(REXML::Document.new(filename)).to eq(REXML::Document.new(expected))` but doesn't work. What did you have in mind?

Reply:
In the end, I used file.read and it worked. Another option would be to check the result partially like: